content/en/blog/_posts/2023-05-08-volume-group-snapshot-alpha.md

```diff
@@ -0,0 +1,268 @@
+spec:
+  driver: com.example.csi-driver
+  deletionPolicy: Delete
+  source:
+     volumeGroupSnapshotHandle: group-snap-id
+  volumeGroupSnapshotRef:
+    kind: VolumeGroupSnapshot
+    name: pre-existing-group-snap1
+    namespace: demo-namespace
+```
+
+A VolumeGroupSnapshot object should be created to allow a user to use the group snapshot:
+
+```yaml
+apiVersion: groupsnapshot.storage.k8s.io/v1alpha1
+kind: VolumeGroupSnapshot
+metadata:
+  name: pre-existing-group-snap1
+  namespace: demo-namespace
+spec:
+  snapshotContentName: pre-existing-group-snap-content1
+```
+
+Once these objects are created, the snapshot controller will bind them together,
+and set the field `status.ready` to `"True"` to indicate the group snapshot is ready
+to use.
+
+### How to use group snapshot for restore in Kubernetes
+
+At restore time, the user can request a new PersistentVolumeClaim to be created from
+a VolumeSnapshot object that is part of a VolumeGroupSnapshot. This will trigger
+provisioning of a new volume that is pre-populated with data from the specified
+snapshot. The user should repeat this until all volumes are created from all the
+snapshots that are part of a group snapshot.
+
+## As a storage vendor, how do I add support for group snapshots to my CSI driver?
+
+To implement the volume group snapshot feature, a CSI driver MUST:
+
+* Implement a new group controller service.
+* Implement group controller RPCs: `CreateVolumeGroupSnapshot`, `DeleteVolumeGroupSnapshot`, and `GetVolumeGroupSnapshot`.
+* Add group controller capability `CREATE_DELETE_GET_VOLUME_GROUP_SNAPSHOT`.
+
+See the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)
+and the [Kubernetes-CSI Driver Developer Guide](https://kubernetes-csi.github.io/docs/)
+for more details.
+
+Although Kubernetes poses as little prescriptive on the packaging and deployment of
+a CSI Volume Driver as possible, it provides a suggested mechanism to deploy a
+containerized CSI driver to simplify the process.
+
+As part of this recommended deployment process, the Kubernetes team provides a number of
+sidecar (helper) containers, including the
+[external-snapshotter sidecar container](https://kubernetes-csi.github.io/docs/external-snapshotter.html)
+which has been updated to support volume group snapshot.
+
+The external-snapshotter watches the Kubernetes API server for the
+`VolumeGroupSnapshotContent` object and triggers `CreateVolumeGroupSnapshot` and
+`DeleteVolumeGroupSnapshot` operations against a CSI endpoint.
+
+## What are the limitations?
+
+The alpha implementation of volume group snapshots for Kubernetes has the following
+limitations:
+
+* Does not support reverting an existing PVC to an earlier state represented by a snapshot that is part of a group snapshot (only supports provisioning a new volume from a snapshot).
+* No application consistency guarantees beyond any guarantees provided by the storage system (e.g. crash consistency).
+
```

Comment:
What about adding this?
```suggestion

Whether you plan to quiesce your application or take a live snapshot, you should consider
whether to request a freeze for any filesystem backed by block storage. For example, use
`ioctl(fd, FIFREEZE, 0);` to trigger a freeze, then take the (group) snapshot, and finally
trigger a thaw.

```

I don't know if the fzfreeze is a responsibility of the CSI Driver, but I assume that it _isn't_.

Reply:
fsfreeze is not done by the CSI driver. There are some discussions about this here: https://github.com/container-storage-interface/spec/pull/407. 
There's also this effort here: https://github.com/kubernetes/enhancements/pull/1995

I'm hesitant to give an example using fsfreeze because application quiesce could mean using database specific methods, i.e., db.fsyncLock() for MongoDB.

Reply:
The point I want to make is that you might incorrectly think that `fsync()` is enough. For some filesystems it might be. For others, the lack of a `FIFREEZE` might mean the difference between a snapshot that is recoverable and a snapshot that probably will be but we're not sure.

We _don't_ need to mention the ioctl but we might want to let readers know that there's more to it than quiescing at app level and hoping. A particularly unhelpful setup might do a heartbeat write every second even when the app isn't active, for example. And then that gets split over multiple blocks at the storage layer. I have a feeling I've even worked with a system with these properties.

Reply:
Sure.  I added a reference to this doc: https://github.com/kubernetes/community/blob/master/wg-data-protection/data-protection-workflows-white-paper.md#quiesce-and-unquiesce-hooks